### DISCUSSION TITLE
Intercepting key events regardless of focus

### DISCUSSION THREAD

**Discussion by etiennep249** [slint-ui/slint#4231] [Q&A]
I'd like to be able to intercept key events without a FocusScope. With a LineEdit, for example, it steals the focus from whatever might have a FocusScope, so you can't handle keybinds while in it. Or have a "global" keybind that runs regardless of where focus might be.

I'm using winit, which has key events. However, with slint, we don't have access to this event loop and building another one is forbidden. There are some crates out there that support global keybinds, but I could not find one that supports both X11 and Wayland like winit does.

Is there something that would help me achieve this?

**Comment by hunger** [comment]
If you know that you are using Winit, then you could try to use the `i-slint-nackend-winit` crate (internal, so no stability guarantees!). See the [docs](https://docs.rs/i-slint-backend-winit/latest/i_slint_backend_winit/) for a way to access the winit window directly from there.

I have not tried this, but maybe you can intercept the events there :-)

**Reply by etiennep249** [reply]
I tried it, it doesn't give any access to the event loop (that I am aware of). It seems to be more to interact with the window itself.
Looking at [event_loop.rs](https://github.com/slint-ui/slint/blob/master/internal/backends/winit/event_loop.rs), there doesn't seem to be any external way to access it.  The interface only contains the window, the proxy to send events and the clipboard. Worst case scenario, I could fork the whole crate just to be able to pass a callback to the event loop, but that seems a bit extreme.

**Comment by ogoffart** [comment]
I wonder if, in focus scope, we shouldn't maybe have a new `intercept-key-pressed(KeyEvent) -> EventResult`
It would work exactly like `key-pressed` but instead of sending event from top to bottom of the focus scope, we would send them from bottom to top, before the `key-pressed`.  This would allow to capture shortcut.

```
FocusScope {
   intercept-key-pressed(event) => {  
      ... if the event is CTRL + C , send the copy() callback  and accept..
   }
   
  
   ...
        LineEdit {
        }
}
```

So if user press CTRL + C, then we save and nothing is sent to the line edit (so it override the LineEdit shortcut) otherwise, for other keys, the LineEdit keeps getting all the shortcut.

@tronical: what do you think?

**Reply by tronical** [reply]
Hmm, so how would "bottom to top" look like? We traverse top to bottom, collect all scopes on the way, and then invoke `intercept-key-pressed` in reverse order?

That said, a global key bind regardless of where the focus is, that's in my opinion a shortcut. As the name "global" implies, for the user this should always work if the _window_ is active. I would not add tree traversal to this global phenomenon but rather have a `Shortcut` element, for example.

Say you have a "Ctrl+C" shortcut that should copy something depending on the rest of the application state, then I'd rather have that code in one place ("if (this_mode) copy_that() else if (that_mode) copy_something_else()") instead of relying on a reverse tree traversal that may be much harder to debug.

**Reply by bitmural** [reply]
Has this concept been implemented anywhere yet in Slint? I love global a action or search pane where a static keybind performed in any context opens a nice temporary interface, just like ctrl + k on the docs site. I haven't seen the way to implement such a feature using the existing FocusScope and forward-focus on the Window (or toolbar).

**Reply by ogoffart** [reply]
Yes, this has been implemented in FocusScope::capture-key-pressed https://docs.slint.dev/latest/docs/slint/reference/keyboard-input/focusscope/#capture-key-pressedkeyevent---eventresult

**Reply by bitmural** [reply]
I had thought that was supposed to be it, but I'm experiencing that the standard widgets will prevent the window level focus scope from handling the input event. Here is the example scenario, where moving focus around with TAB will not trigger a message from debug:

```slint
import { Button, CheckBox, LineEdit, HorizontalBox, VerticalBox } from "std-widgets.slint";

export component AppWindow inherits Window {
  width: 400px; height: 300px;
  forward-focus: globalFocus;
  globalFocus := FocusScope {
    capture-key-pressed(event) => {
      if (event.modifiers.control && event.text == "s") {
        debug("Global 'Save' Hotkey Pressed!");
        return EventResult.accept;
      } else { return EventResult.reject; }
    }
  }
  VerticalLayout {
    alignment: LayoutAlignment.space-between;
    VerticalBox {
      HorizontalBox {
        preferred-height: 28px;
        Text {
          vertical-alignment: TextVerticalAlignment.center;
          text: "Record";
        }
        LineEdit { }
      }
      CheckBox { text: "I'm a check box!"; }
    }
    HorizontalBox {
      preferred-height: 28px;
      Button { text: "Save"; }
    }
  }
}
```

However, I did just stumble across this recent blog post, [Making Slint Desktop Ready](https://slint.dev/blog/making-slint-desktop-ready), discussing the exact features I had been curious to try in a small application. The post includes global shortcuts.

I will keep a keen eye on the issues noted in the post for future use! Thanks
